### PR TITLE
vmselect: add query tracing feature

### PR DESCRIPTION
**Pull request by hagen1778**
Query tracing feature allows displaying the stages of query processing.
The stages can be then visualized in order to understand the slowest
or the most expensive parts of the query processing.

The tracing is implemented via `lib/tracer`. The library allows creating
a tracing context, which supports nesting. The context is then used
in `vmselect` and `lib/storage` in order to highlight the most interesting
parts of query processing.

The query tracing can be enabled for each specific query by passing extra
GET params in request. The trace will be then collected and printed to
the stdout and also attached to the response (only `/api/v1/query` and
`api/v1/query_range` support tracing in response).

See more at `tracer/example_context_test.go`

Signed-off-by: hagen1778 <roman@victoriametrics.com>

-----------------------

Example:

```
curl 'http://localhost:8428/prometheus/api/v1/query_range?query=(sum(sum_over_time(%7Bjob%3D%22avalanche%22%2C%20__name__%3D~%22avalanche_metric_mmmmm_0_1.*%22%7D%5B2d%5D))%20%2B%20sum(count_over_time(%7Bjob%3D%22avalanche%
22%2C%20__name__%3D~%22avalanche_metric_mmmmm_0.*%22%7D%5B10d%5D)))%20&start=1648795412&end=1651563812.851&step=22147.207&nocache=1&trace_query=true'

 2314ms: query trace "16EB9FDB246980D9"
-- 2314ms: Exec on range "769h0m0.851s" with step 22147s returned 1 series 125 samples
---- 2313ms: evaluating expression "sum(sum_over_time({job=\"avalanche\", __name__=~\"avalanche_metric_mmmmm_0_1.*\"}[2d])) + sum(count_over_time({job=\"avalanche\", __name__=~\"avalanche_metric_mmmmm_0.*\"}[10d]))"
------ 237ms: evaluating expression "sum(sum_over_time({job=\"avalanche\", __name__=~\"avalanche_metric_mmmmm_0_1.*\"}[2d]))"
-------- 237ms: evalRollupFuncWithMetricExpr 8 time series, 125 points each, total 1000 samples at once, with estimated memory 16000 bytes
---------- 62ms: index search on time range "817h5m0.851s" returns estimation of 12220 series
---------- 117ms: ProcessSearchQuery scanned 116850 blocks, 12220 series, 4204706 samples
---------- 51ms: RunParallel processed 12220 series, 4204706 rows via 8 workers
------ 2075ms: evaluating expression "sum(count_over_time({job=\"avalanche\", __name__=~\"avalanche_metric_mmmmm_0.*\"}[10d]))"
-------- 2075ms: evalRollupFuncWithMetricExpr 8 time series, 125 points each, total 1000 samples at once, with estimated memory 16000 bytes
---------- 125ms: index search on time range "1009h5m0.851s" returns estimation of 110000 series
---------- 1409ms: ProcessSearchQuery scanned 1098485 blocks, 110000 series, 34435475 samples
---------- 498ms: RunParallel processed 110000 series, 34435475 rows via 8 workers
------ 0ms: binary operation "+" returned 1 series
-- 0ms: Exec transforms 1 series to result of 1 series
```

### CODE REVIEW

File: lib/tracer/context.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	id string
+	// msg is the message/title of the current context
+	msg string
+	// start is the time when Context was created
+	start time.Time
+	// enabled makes all function calls no-op if set to false
+	enabled bool
+	// color defines whether to print tracing with colors in stdout
+	color bool
+
```
Comment (f41gh7) [inline lib/tracer/context.go:77]: I think, that optional key=value pairs may be useful for context. 

It simplifies programmatic trace analyze.

 For instance, i want to group traces by top processed rows. Or by top executed function.

 Something like:
type KeyValue struct{
   Key string
   Value interface {}
}
```
func (c *Context) Done(msg string, kv KeyValue...){}...
...

func someQuery(){
  ctx.Done("processed parallel query", {"ts_len", 124124}, {"function", "up"})
}

```

File: app/vmselect/main.go [hunk 2]
```diff
@@ -119,6 +120,14 @@ func RequestHandler(w http.ResponseWriter, r *http.Request) bool {
 		}
 	}
 
+	ctx := tracer.NewContext(r)
+	defer func() {
+		ctx.Done(func() string {
+			return fmt.Sprintf("query trace %q", ctx.ID())
+		})
+		fmt.Println(ctx.Print())
```
Comment (f41gh7) [inline app/vmselect/main.go:128]: I'm not sure, that Println must be used, it breaks logs parsing. Is it possible to use standard logger?

**Review by f41gh7 — APPROVED** [review]
LGTM, thanks!

 have few comments

**Review by dmitryk-dk — APPROVED** [review]
LGTM!

**Comment by valyala** [comment]
Manually merged this pull request in the commit 41958ed5dd57e779b63d12c5319afa73e2681561 . @hagen1778 , thanks for the initial implementation.

**Pull request closed without merging**